pkg/operator/resource/resourceapply/monitoring.go

```diff
@@ -2,148 +2,159 @@ package resourceapply
 	if err != nil {
-		recorder.Warningf("ServiceMonitorUpdateFailed", "Failed to update ServiceMonitor.monitoring.coreos.com/v1: %v", err)
-		return nil, true, err
+		return nil, false, err
 	}
-
-	recorder.Eventf("ServiceMonitorUpdated", "Updated ServiceMonitor.monitoring.coreos.com/v1 because it changed")
-	return newObj, true, err
-}
-
-var prometheusRuleGVR = schema.GroupVersionResource{Group: "monitoring.coreos.com", Version: "v1", Resource: "prometheusrules"}
-
-// ApplyPrometheusRule applies the PrometheusRule
-func ApplyPrometheusRule(ctx context.Context, client dynamic.Interface, recorder events.Recorder, required *unstructured.Unstructured) (*unstructured.Unstructured, bool, error) {
-	namespace := required.GetNamespace()
-
-	existing, err := client.Resource(prometheusRuleGVR).Namespace(namespace).Get(ctx, required.GetName(), metav1.GetOptions{})
-	if errors.IsNotFound(err) {
-		newObj, createErr := client.Resource(prometheusRuleGVR).Namespace(namespace).Create(ctx, required, metav1.CreateOptions{})
-		if createErr != nil {
-			recorder.Warningf("PrometheusRuleCreateFailed", "Failed to create PrometheusRule.monitoring.coreos.com/v1: %v", createErr)
-			return nil, true, createErr
-		}
-		recorder.Eventf("PrometheusRuleCreated", "Created PrometheusRule.monitoring.coreos.com/v1 because it was missing")
-		return newObj, true, nil
+	if !found {
+		return nil, false, errorsstdlib.New(fmt.Sprintf("metadata not found in %s", required.GetName()))
 	}
+
+	// Cast the metadata to the correct type.
+	var existingObjectMetaTyped, requiredObjectMetaTyped metav1.ObjectMeta
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(existingObjectMeta, &existingObjectMetaTyped)
 	if err != nil {
 		return nil, false, err
 	}
-
-	existingCopy := existing.DeepCopy()
-
-	toUpdate, modified, err := ensureGenericSpec(required, existingCopy, noDefaulting, equality.Semantic)
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(requiredObjectMeta, &requiredObjectMetaTyped)
 	if err != nil {
 		return nil, false, err
 	}
 
-	if !modified {
-		return nil, false, nil
-	}
-
-	if klog.V(2).Enabled() {
-		klog.Infof("PrometheusRule %q changes: %v", namespace+"/"+required.GetName(), JSONPatchNoError(existing, toUpdate))
-	}
-
-	newObj, err := client.Resource(prometheusRuleGVR).Namespace(namespace).Update(ctx, toUpdate, metav1.UpdateOptions{})
+	// Deep-check the spec objects for equality, and update the cache in either case.
+	existingCopy, modified, err := ensureGenericSpec(required, existingCopy, noDefaulting, equality.Semantic)
 	if err != nil {
-		recorder.Warningf("PrometheusRuleUpdateFailed", "Failed to update PrometheusRule.monitoring.coreos.com/v1: %v", err)
-		return nil, true, err
+		return nil, false, err
 	}
-
-	recorder.Eventf("PrometheusRuleUpdated", "Updated PrometheusRule.monitoring.coreos.com/v1 because it changed")
-	return newObj, true, err
+	if !modified {
```

Comment:
If spec did not change, but labels do change.  How does this code avoid returning early?

Reply:
Addressed in the latest commit, also added some tests.